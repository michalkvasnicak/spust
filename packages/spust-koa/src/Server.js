// @flow

import Koa from 'koa';
import PropTypes from 'prop-types';
import React from 'react';
import { Server as HttpServer } from 'http';

// eslint-disable-next-line no-unused-vars
type ExtractMiddlewareType<MT: *, T: (mw: MT) => *> = MT;

export type MiddlewareFn = ExtractMiddlewareType<any, $PropertyType<Koa, 'use'>>;

export type ServerContext = {
  app: Koa,
  listen: (...args: any) => HttpServer,
};

export type Props = {
  context: Object,
  children?: any,
  port: number,
};

type DefaultProps = {
  context: Object,
};

export type Context = {
  server: Koa,
  use: $PropertyType<Koa, 'use'>,
};

export const serverContextType = {
  use: PropTypes.func.isRequired,
  server: PropTypes.instanceOf(Koa).isRequired,
};

export default class Server extends React.Component<DefaultProps, Props, void> {
  app: Koa;

  static defaultProps = {
    context: {},
  };

  static childContextTypes = serverContextType;

  constructor(props: Props, context: any) {
    super(props, context);

    this.app = new Koa();

    this.props.context.app = this.app;
    this.props.context.listen = () => this.app.listen(this.props.port);
  }

  getChildContext = () => ({
    use: (middleware: MiddlewareFn) => this.app.use(middleware),
    server: this.app,
  });

  render() {
    return (
      <div>
        {this.props.children}
      </div>
    );
  }
}
