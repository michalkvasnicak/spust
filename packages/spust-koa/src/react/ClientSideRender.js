// @flow

import { assets as loadAssets } from 'spust';
import React from 'react';

import RenderApp from '../RenderApp';

type Props = {
  containerId?: string,
  lang?: string,
  title?: string,
};

function createRenderer(props: Props) {
  let assets;
  let scripts = [];
  let stylesheets = [];

  return ctx => {
    // load assets everytime during development mode
    if (assets == null || process.env.NODE_ENV !== 'production') {
      assets = loadAssets();
      scripts = assets.js.map(src => ({ src }));
      stylesheets = assets.css.map(href => ({ href }));
    }

    return {
      containerId: props.containerId,
      head: {
        title: props.title,
      },
      lang: props.lang,
      scripts,
      stylesheets,
    };
  };
}

export default function ClientSideRender(props: Props) {
  return <RenderApp render={createRenderer(props)} />;
}
