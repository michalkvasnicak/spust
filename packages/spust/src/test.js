#!/usr/bin/env node
// @flow

process.env.NODE_ENV = 'test';
process.env.BABEL_ENV = 'test';

process.on('unhandledRejection', err => {
  throw err;
});

const jest = require('jest');
const fs = require('fs');
const path = require('path');
const argv = process.argv.slice(2);

if (!process.env.CI && argv.indexOf('--coverage') < 0) {
  argv.push('--watch');
} else {
  argv.push('--forceExit');
}

const setupTestsFile = fs.existsSync(path.resolve(process.cwd(), 'src/setupTests.js'))
  ? '<rootDir>/src/setupTests.js'
  : undefined;

argv.push(
  '--config',
  JSON.stringify({
    collectCoverageFrom: ['src/**/*.{js,jsx}'],
    setupFiles: [path.resolve(__dirname, './polyfills/server.js')],
    setupTestFrameworkScriptFile: setupTestsFile,
    testMatch: [
      '<rootDir>/src/**/__tests__/**/*.js?(x)',
      '<rootDir>/src/**/?(*.)(spec|test).js?(x)',
    ],
    testEnvironment: 'node',
    testURL: 'http://localhost',
    transform: {
      '^.+\\.(js|jsx)$': path.resolve(__dirname, 'jest/babelTransform.js'),
      '^.+\\.css$': path.resolve(__dirname, 'jest/cssTransform.js'),
      '^(?!.*\\.(js|jsx|css|json)$)': path.resolve(__dirname, 'jest/fileTransform.js'),
    },
    transformIgnorePatterns: ['[/\\\\]node_modules[/\\\\].+\\.(js|jsx)$'],
  }),
);

jest.run(argv);
