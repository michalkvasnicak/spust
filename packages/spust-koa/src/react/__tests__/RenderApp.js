// @flow

import path from 'path';
import React from 'react';
import test from 'supertest';

import ClientSideRender from '../ClientSideRender';
import Server from '../../Server';
import serve from '../../';

// assets function from Spust is working with this internally
// provide valid path, in your case see assets.json
process.env.__ASSETS_JSON_PATH = path.resolve(__dirname, './assets.json');

describe('RenderApp', () => {
  it('works correctly', async () => {
    const server = serve(
      <Server port={3000}>
        <ClientSideRender containerId="root" title="Test title" />
      </Server>,
      false,
    );

    await test(server.app.callback())
      .options('/')
      .expect(200)
      .expect(res => expect(res.text).toMatchSnapshot());
  });
});
