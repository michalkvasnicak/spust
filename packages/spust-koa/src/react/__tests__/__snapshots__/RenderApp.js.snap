// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderApp works correctly 1`] = `
"<html lang=\\"en\\">
  <head><title>Test title</title></head>
  <body>
    <div id=\\"root\\"></div>
    <script src=\\"/static/js/main.d6c4cf02.js\\"></script>
  </body>
</html>"
`;
